Size tests

DennisLfromGA opened the pull request:
Per Issue https://github.com/dnschneid/crouton/issues/2108 - Add destination size tests:
If the CHROOTS partition size is less than 4 GB, spit out a warning. If CHROOTS free space is less than 2GB, spit out a warning. Also, determine whether the former is due to a Chrubuntu install.

DennisLfromGA committed:
Merge pull request #1 from dnschneid/master

Updating fork...

DennisLfromGA committed:
calls to check_space and check_chrubuntu functions

Add routine to check space on CHROOTS partiton and ChrUbuntu partitions

DennisLfromGA committed:
add functions: check_space and check_chrubuntu

check_space checks that the minimum partition size and free space is met.
check_ubuntu checks if partition 7 is being used and reports it's sizes and configuration

DennisLfromGA committed:
added return values for check_chrubuntu

Added return values for check_chrubuntu in case we decide to use them.

DennisLfromGA committed:
updated 'df' command in check_chrubuntu to use mountpoint

'df' command in check_chrubuntu changed to use mountpoint instead of partition.

DennisLfromGA committed:
moved check_space variable declarations

Moved check_space variable declarations to top of file with the others.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+# $1: the partition's minimum size threshold
+# $2: the partition's minimum available space threshold
+check_space() {
+    local part=${CHROOTS:-/usr/local}
```
quoting

DennisLfromGA replied:
quotes added.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+check_space() {
+    local part=${CHROOTS:-/usr/local}
+    df -m $part | awk 'FNR == 2 {print $2,$4,$6}' |\
+    while read -r size avail mp; do
```
This doesn't need to be a while loop, since you always "return" after one round.  Just start a `{ }` block after the `|` on the upper line.

DennisLfromGA replied:
while loop changed to `{ }` block.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+        if [ "$size" -lt "$1" ]; then
+            local size_gb="$((${size}/1000))GB"
+            echo "WARNING: Your CHROOTS partition is rather small (${size_gb})."
+            echo "WARNING: Usually $((${1}/1000))GB or more is needed. "
```
"XGB or more is recommended."

DennisLfromGA replied:
re-worded WARNING.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+            local size_gb="$((${size}/1000))GB"
+            echo "WARNING: Your CHROOTS partition is rather small (${size_gb})."
+            echo "WARNING: Usually $((${1}/1000))GB or more is needed. "
+        #   error 1 "ERROR: Choose another partition then try again..."
```
remove the dead lines, please

DennisLfromGA replied:
commented lines removed.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+    done
+}
+
+# Check if partition 7 / ROOT-C (ChrUbuntu) has been resized
```
Seeing how ChrUbuntu is mostly being superseded by other projects, it's probably worth changing references to just "dual boot"

DennisLfromGA replied:
renamed function and all variable names from 'chrubuntu' to 'dualboot' and updated comments and messages..

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+# Check if partition 7 / ROOT-C (ChrUbuntu) has been resized
+# normally called when space requirements on the CHROOTS partiton aren't met
+check_chrubuntu() {
+    local disk=$(rootdev -s -d)
```
quoting (also lines below)

DennisLfromGA replied:
quotes added.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+    local chrubuntu_part=$(rootdev -s | sed 's/.$/7/')
+    local chrubuntu_size=$(cgpt show -i7 -s ${disk})
+    if [ "$chrubuntu_size" -gt 1 ]; then
+        echo "WARNING: Partiton 7, normally used for ChrUbuntu, has been resized"
```
"often used for dual-boot installs"

DennisLfromGA replied:
reworded and changed 'chrubuntu' to 'dual-boot installs'.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+        echo "WARNING: Partiton 7, normally used for ChrUbuntu, has been resized"
+        local tmp="`mktemp -d --tmpdir=/tmp 'check-chrubuntu.XXX'`"
+        local unmount="umount -l '$tmp' && rmdir '$tmp'"
+    #   addtrap "$unmount"
```
Why not?

DennisLfromGA replied:
trap added and later undone.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+    #   addtrap "$unmount"
+        mount $chrubuntu_part "$tmp" 2>/dev/null
+        df -m "$tmp" | awk 'FNR == 2 {print $2,$3,$4}' |\
+        while read -r size used avail; do
```
This also could probably just be a `{ }` block

DennisLfromGA replied:
while loop changed to `{ }` block.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+            local chrubuntu_size_gb="$((${size}/1000))GB"
+            local chrubuntu_used_gb="$((${used}/1000))GB"
+            local chrubuntu_avail_gb="$((${avail}/1000))GB"
+            echo "WARNING: to ${chrubuntu_size_gb}, it's using ${chrubuntu_used_gb}, and ${chrubuntu_avail_gb} is available for your use..."
```
`to ${chrubuntu_size_gb}, with ${chrubuntu_avail_gb} free`

DennisLfromGA replied:
re-worded per note.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,60 @@ websocketcommand() {
+        done
+        distrib="$(grep 'DISTRIB_DESCRIPTION=' "$tmp/etc/lsb-release" 2>/dev/null | cut -d= -f2)"
+        if [ -n "$distrib" ]; then
+            echo "WARNING: Also, it has $distrib installed."
```
`(It appears to have $distrib installed.)`

DennisLfromGA replied:
re-worded per note.

DennisLfromGA replied:
trap added above and undone here.

dnschneid commented on installer/main.sh:
```diff
@@ -366,6 +368,15 @@ addtrap "stty echo 2>/dev/null"
 CHROOTS="$PREFIX/chroots"
 
+# Check if space requirements have been met
+# if not, check for chrubuntu
```
clean up comments

DennisLfromGA replied:
re-worded comments

dnschneid commented on installer/main.sh:
```diff
@@ -366,6 +368,15 @@ addtrap "stty echo 2>/dev/null"
+#PART_SIZE_MIN=400000  # in MB !!! TEST VALUE
+if ! check_space "$PART_SIZE_MIN" "$PART_AVAIL_MIN"; then
+    check_chrubuntu
+#   exit 1 # Don't exit, just spit out out warnings
```
probably do the same wait-for-x-seconds thing we do elsewhere

DennisLfromGA replied:
sleep X added to both checks.

dnschneid commented:
Thanks for the patch!!  Catching up on reviewing things now :)

DennisLfromGA committed:
updated per line notes

Updated to incorporate changes per line notes from @dnschneid

DennisLfromGA committed:
updated per line notes

Updated to incorporate changes per line notes from @dnschneid

DennisLfromGA commented:
> Thanks for the patch!! Catching up on reviewing things now :)

Thank you for all the help and time out of your busy schedule. :)

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,58 @@ websocketcommand() {
+check_space() {
+    local part="${CHROOTS:-/usr/local}"
+    df -m $part | awk 'FNR == 2 {print $2,$4}' | {
+    read -r size avail
```
indentation

DennisLfromGA replied:
done.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,58 @@ websocketcommand() {
+        addtrap "$unmount"
+        mount $dualboot_part "$tmp" 2>/dev/null
+        df -m "$tmp" | awk 'FNR == 2 {print $2,$3,$4}' | {
+        read -r size used avail
```
indentation

DennisLfromGA replied:
done.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,58 @@ websocketcommand() {
+        df -m "$tmp" | awk 'FNR == 2 {print $2,$3,$4}' | {
+        read -r size used avail
+            local dualboot_size_gb="$((${size}/1000))GB"
+            local dualboot_used_gb="$((${used}/1000))GB"
```
this isn't used

DennisLfromGA replied:
removed 'used' from awk output and elsewhere.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,58 @@ websocketcommand() {
+        local tmp="`mktemp -d --tmpdir=/tmp 'check-dualboot.XXX'`"
+        local unmount="umount -l '$tmp' 2>/dev/null && rmdir '$tmp'"
+        addtrap "$unmount"
+        mount $dualboot_part "$tmp" 2>/dev/null
```
`mount -o ro`

dnschneid replied:
also, quoting on $dualboot_part

DennisLfromGA replied:
added 'ro' option and quoted.

DennisLfromGA committed:
installer/functions updated per line notes

Updated to incorporate changes per line notes from @dnschneid on 12/30/15.

DennisLfromGA committed:
installer/functions updated to remove 'used'.

Updated to remove remnant of 'used' on line 280.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+# $2: the partition's minimum available space threshold
+check_space() {
+    local part="${CHROOTS:-/usr/local}"
+    df -m $part | awk 'FNR == 2 {print $2,$4}' | {
```
quote $part

DennisLfromGA replied:
quoted.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+                echo "WARNING: Usually $((${2}/1000))GB or more is recommended. "
+                return 2
+            else
+                echo "Your space requirements for CHROOTS have been met."
```
This output isn't really necessary

DennisLfromGA replied:
removed superfluous message.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+    local part="${CHROOTS:-/usr/local}"
+    df -m $part | awk 'FNR == 2 {print $2,$4}' | {
+        read -r size avail
+            if [ "$size" -lt "$1" ]; then
```
whoa, too much indentation :)
read and if are at the same level.

DennisLfromGA replied:
fixed indentation.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+        mount -o ro "$dualboot_part" "$tmp" 2>/dev/null
+        df -m "$tmp" | awk 'FNR == 2 {print $2,$4}' | {
+            read -r size avail
+                local dualboot_size_gb="$((${size}/1000))GB"
```
same indentation issue here

DennisLfromGA replied:
fixed indentation.
also reworked logic to check for existing mount.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+check_dualboot() {
+    local disk="$(rootdev -s -d)"
+    local dualboot_part="$(rootdev -s | sed 's/.$/7/')"
+    local dualboot_size="$(cgpt show -i7 -s ${disk})"
```
probably should quote $disk

DennisLfromGA replied:
quoted.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+                local dualboot_avail_gb="$((${avail}/1000))GB"
+                echo "WARNING: to ${dualboot_size_gb}, with ${dualboot_avail_gb} free."
+        }
+        distrib="$(grep 'DISTRIB_DESCRIPTION=' "$tmp/etc/lsb-release" 2>/dev/null | cut -d= -f2)"
```
once you fix the indentation, please make sure all lines are <80 chars wide

DennisLfromGA replied:
adjusted line lengths to <80chars 'ish...

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+        read -r size avail
+            if [ "$size" -lt "$1" ]; then
+                local size_gb="$((${size}/1000))GB"
+                echo "WARNING: Your CHROOTS partition is rather small (${size_gb})."
```
I think we've been outputting warnings like these out to stderr, but might want to check for consistency

DennisLfromGA replied:
warnings sent to stderr.

dnschneid commented on installer/functions:
```diff
@@ -239,3 +239,57 @@ websocketcommand() {
+            if [ "$size" -lt "$1" ]; then
+                local size_gb="$((${size}/1000))GB"
+                echo "WARNING: Your CHROOTS partition is rather small (${size_gb})."
+                echo "WARNING: Usually $((${1}/1000))GB or more is recommended. "
```
extra space at end of string

DennisLfromGA replied:
fixed.

dnschneid commented on installer/main.sh:
```diff
@@ -366,6 +368,15 @@ addtrap "stty echo 2>/dev/null"
+# Check if space requirements have been met
+# if not, check for dualboot
+if ! check_space "$SIZEMIN" "$AVAILMIN"; then
+    sleep 3
```
probably don't want output, delay, more output, more delay.

the other instances we do this, we have extra text that says something like "installation will continue in 5 seconds, or hit ctrl-c to cancel"...copy that text?

DennisLfromGA replied:
Removed multiple delays and added message that installation will continue after 5s.

DennisLfromGA committed:
installer/main.sh updated per line notes

Updated to incorporate changes per line notes from @dnschneid on 1/1/16.
Removed multiple delays and added message that installation will continue after 5s.

DennisLfromGA committed:
installer//functions updated per line notes

Updated to incorporate changes per line notes from @dnschneid on 1/1/16.
Added missing quotes, fixed indentation, insured lines <=80 chars, sent warnings to stderr.
Also reworked logic to check for existing mounts do to 'ro' mount failing if already mounted.

DennisLfromGA commented:
I apologize for all the misstarts ( I'm trying, very trying ;).

dnschneid commented:
You're doing great!

DennisLfromGA committed:
installer/main.sh updated per line notes on 1/4/2015

Simplified logic for invoking check_dualboot function.

DennisLfromGA committed:
installer/functions updated per line notes: 1/4/2015

un-split warning messages.
removed `proc/mounts` lookup and redid mount logic for mounted and unmounted partition.
'grep' and 'cut' replaced with 'awk'.

DennisLfromGA commented:
> You're doing great!

Well I certanly don't think so, it's down right embarrassking!

DennisLfromGA commented:
I was thinking of changing some of my calculations to use decimals.

The current ones are similar to:
>  local size_gb="$((${size}/1000))GB"

When size is 19588, it yields '19GB'

--------
I'd like to change them to:
>  local size_gb="$(echo "$size 1000" | awk '{printf "%.2f\n", $1/$2}') GB"

When size is 19588, it yields '19.58 GB'

I wanted to check with y'all and see if that's something worthwhile doing ???

tedm commented:
I would just round to 20GB It's not like you can buy a 19.58GB storage device ?? ;)

DennisLfromGA commented:
Okay, thanx. I'll leave them as is then.

dnschneid commented:
If you want it to round up you can do something like:
`local size_gb="$(((${size}+500)/1000))GB"`

I'm not sure which is preferable...

zxvv commented:
You could also consider using "df -mh"  which will give you the size in 'human' readable format, such as 20GB.

DennisLfromGA commented:
Yup, thought about that but it's just a little harder to test sizes when in a 'human' readable format.
I'm happy with the way it is now unless we want to add the 'echo_color' attributes before it's merged.

DennisLfromGA commented:
I think this is ready to go but you might want to review it again, it's been a while.

tedm commented:
@DennisLfromGA Let me know if you want me to test. During some wild attempts to get my Chromebook (Toshiba 2 2015) to run android apps, I lost Crouton, so have to reinstall, probably this weekend.

DennisLfromGA commented:
@tedm, Go for it and let us know, thanx.
